TLSv1.3 draft 22

mattcaswell opened the pull request:
If draft-22 existed (which it doesn't) then it might look something like this.

This implements the middlebox compat workarounds currently being discussed on the TLS WG email list.

This negotiates as the draft-22 version (0x7f16) on the assumption that that is what this will turn into.

FdaSilvaYY commented on include/openssl/sslerr.h:
```diff
@@ -303,6 +303,7 @@ int ERR_load_SSL_strings(void);
 # define SSL_F_TLS_CONSTRUCT_STOC_SESSION_TICKET          460
 # define SSL_F_TLS_CONSTRUCT_STOC_STATUS_REQUEST          461
 # define SSL_F_TLS_CONSTRUCT_STOC_SUPPORTED_GROUPS        544
+# define SSL_F_TLS_CONSTRUCT_STOC_SUPPORTED_VERSIONS      548
```
```
$ if $make update; then echo -e '+\057 MAKE UPDATE OK'; else echo -e '+\057 MAKE UPDATE FAILED'; false; fi; git diff --exit-code
SSL function code 548 collision at SSL_F_TLS_CONSTRUCT_STOC_SUPPORTED_VERSIONS
make: *** [errors] Error 25
+/ MAKE UPDATE FAILED
The command "if $make update; then echo -e '+/ MAKE UPDATE OK'; else echo -e '+/ MAKE UPDATE FAILED'; false; fi; git diff --exit-code" exited with 0.
```

mattcaswell replied:
Strange...I don't see this collision?

riking replied:
548 used by SSL_F_TLS_CONSTRUCT_STOC_MAXFRAGMENTLEN

FdaSilvaYY replied:
I just copy-paste a piece of a Travis failed jobs.
As I say in #1008, the reworked `util/mkerr.pl` script don't break (all) the build(s) after a `make update` when there is some duplicated IDs .

FdaSilvaYY replied:
Please look for "make update" in this log, for example:
https://travis-ci.org/openssl/openssl/jobs/299726329

mattcaswell replied:
That is very interesting and highlights something about the way that Travis works that I had not appreciated. The value `SSL_F_TLS_CONSTRUCT_STOC_MAXFRAGMENTLEN` does not exist and there is no conflict *in this branch*. However it *does* exist in the master branch. It appears to me that travis is doing a merge of this branch with master before it attempts to build. I had always assumed that the build took place on this branch only.

kaduk replied:
Yes, github provides a candidate-merge commit for a pull request, which is why travis refuses to provide any results when there is a merge conflict -- the candidate-merge commit does not exist in a usable state.

mattcaswell commented:
I've updated this to include the necessary HRR changes so this should now meet tlswg/tls13-spec#1091. This still doesn't have any tests, but I'm going to pause work on this for now while discussion is ongoing around the spec.

kaduk commented on ssl/statem/extensions.c:
```diff
@@ -256,11 +257,13 @@ static const EXTENSION_DEFINITION ext_defs[] = {
-        SSL_EXT_CLIENT_HELLO | SSL_EXT_TLS_IMPLEMENTATION_ONLY
-        | SSL_EXT_TLS1_3_ONLY,
+        SSL_EXT_CLIENT_HELLO | SSL_EXT_TLS1_2_SERVER_HELLO
+        | SSL_EXT_TLS1_3_SERVER_HELLO | SSL_EXT_TLS_IMPLEMENTATION_ONLY,
```
Need SSL_EXT_TLS1_3_HELLO_RETRY_REQUEST here, too.

mattcaswell replied:
Fixed

kaduk commented:
I am not sure if this behavior is new to this patch or also exists in master, but when I send "GET / HTTP/0.9" in early data to an apache built against this pull request, early data is accepted and the regular SSL_read() from apache picks up the early data from the client, and apache sends me a 400 Bad Request and closes the connection.  This is exactly what we didn't want to happen -- early data should only be exposed to the application via SSL_read_early_data() and should be discarded if the application does not call SSL_read_early_data().

kaduk commented:
We also need some more changes to handle the client-side of the HelloRetryRequest flow.

mattcaswell commented:
> I am not sure if this behavior is new to this patch or also exists in master, but when I send "GET / HTTP/0.9" in early data to an apache built against this pull request, early data is accepted and the regular SSL_read() from apache picks up the early data from the client, and apache sends me a 400 Bad Request and closes the connection. This is exactly what we didn't want to happen -- early data should only be exposed to the application via SSL_read_early_data() and should be discarded if the application does not call SSL_read_early_data().

Are you able to repeat your test against master?

mattcaswell commented:
> We also need some more changes to handle the client-side of the HelloRetryRequest flow.

Err...like what specifically? I am aware of three issues with this PR at the moment

- HRR does not send supported_versions
- The CCS sent before early_data has the wrong record version number in it
- Server is sending the CCS in the wrong place after an HRR (comes after 2nd SH rather than after 1st)

Are you saying there is something else missing?

mattcaswell commented:
I have rebased (to resolve the error symbol conflict problem) and added commits to fix the above three bugs.

kaduk commented:
> Are you able to repeat your test against master?

It will take some effort but is doable.

> > We also need some more changes to handle the client-side of the HelloRetryRequest flow.
> Err...like what specifically? I am aware of three issues with this PR at the moment

The CCS received after the HelloRetryRequest-in-ServerHello-clothes does not actually engage a change of cipher, and IIUC the client state machine needs to be prepared to accept CCS coming from TLS_ST_CW_CLNT_HELLO, and also to accept SSL3_MT_SERVER_HELLO coming from TLS_ST_CR_CHANGE.

My branch at https://github.com/kaduk/openssl/commits/draft22hrr contains super-hacky code that I was using at the IETF hackathon for interop testing, with a commit to force HRR by sending an empty key_share the first time, and playing whack-a-mole thereafter.

mattcaswell commented:
> The CCS received after the HelloRetryRequest-in-ServerHello-clothes does not actually engage a change of cipher, and IIUC the client state machine needs to be prepared to accept CCS coming from TLS_ST_CW_CLNT_HELLO, and also to accept SSL3_MT_SERVER_HELLO coming from TLS_ST_CR_CHANGE.

This is not correct. This branch should contain everything you need for the client side. Basically all CCS messages received in a TLSv1.3 context are thrown away in the record layer (treated as an empty record) and so they never get presented to the state machine. This seems the most sensible way forward because a CCS can arrive *at any time* during the handshake:

"An implementation may receive an unencrypted record of type change_cipher_spec consisting of the single byte value 0x01 at any time during the handshake and MUST simply drop it without further processing"

mattcaswell commented:
I've rebased this branch to pick up the latest SNI fixes from master that were causing some people interoperability issues.

mattcaswell commented:
I rebased this to bring it in line with the current master

kaduk commented:
The final draft-22 was just issued, BTW.

mattcaswell commented:
> The final draft-22 was just issued, BTW.

Yeah I know....hence the reason I'm dusting this off again :-)

mattcaswell commented:
Updated to add tests. Also fixed a number of bugs identified by the testing and rebased again due to a conflict with master. I've not yet done any interop testing with other implementations, although I know some was done (successfully) by others with earlier iterations of this PR. I think this is sufficiently mature to take it out of WIP - it is ready for review.

mattcaswell commented:
Pushed some fixes for appveyor and travis issues.

kaduk commented:
Travis notes:

```
../test/recipes/90-test_tls13ccs.t ................. 
1..1
Undefined subroutine &main::srctop_file called at ../test/recipes/90-test_tls13ccs.t line 21.
# Looks like your test exited with 255 before it could output anything.
```

kaduk commented on ssl/statem/statem_clnt.c:
```diff
@@ -496,8 +490,17 @@ WRITE_TRAN ossl_statem_client_write_transition(SSL *s)
-        st->hand_state = TLS_ST_CW_CLNT_HELLO;
+    case TLS_ST_CR_SRVR_HELLO:
+        /*
+         * We only get here in TLSv1.3. We just received an HRR, so issue a
```
Only get here in 1.3 because for <=1.2 the server will write something else after ServerHello before the client gets to write?

mattcaswell replied:
Yes

mattcaswell commented on ssl/ssl_locl.h:
```diff
@@ -323,10 +323,14 @@
 
-# define SSL_IS_FIRST_HANDSHAKE(S) ((s)->s3->tmp.finish_md_len == 0)
+# define SSL_IS_FIRST_HANDSHAKE(S) ((s)->s3->tmp.finish_md_len == 0 \
+                                    || (s)->s3->tmp.peer_finish_md_len == 0)
```
The idea here is to make sure we consider a handshake to be the "first" one right up until the last Finished has been sent/received.

kaduk replied:
That part is clear from the diff.  What's not clear from the diff (and involves reading a lot of existing code) is how this changes the behavior of the existing code consuming this macro, and whether any of those behavior changes might be undesirable.

mattcaswell replied:
I don't believe there should be any impact on existing code. I already felt confident about that (having introduced the macro myself earlier this year as part of the TLSv1.3 work), but just to make doubly sure I've just reviewed all existing usage of the macro. I can't see how any of those uses will be adversely impacted by this.

kaduk replied:
Thank you for double-checking!

mattcaswell commented on ssl/ssl_locl.h:
```diff
@@ -1132,6 +1136,12 @@ struct ssl_st {
+     * but is a random value sent in the legacy session id field.
+     */
+    unsigned char tmp_session_id[SSL_MAX_SSL_SESSION_ID_LENGTH];
+    size_t tmp_session_id_len;
```
I actually did start down that route (i.e. reusing the existing session_id) but backed out the changes and went down this route instead. It started having all sorts of subtle consequences and caused breakage. Perhaps it could have been worked through, but in the end I decided that this wasn't actually a session id at all - it is used for a completely different purpose - so reusing the existing session_id field was probably the wrong way to go. It actually adds unnecessary complexity.

kaduk replied:
I'll trust you that it doesn't work out very well, and agree that it is in some sense an object of a semantically different type.

mattcaswell commented on ssl/statem/extensions.c:
```diff
@@ -361,6 +365,49 @@ static int validate_context(SSL *s, unsigned int extctx, unsigned int thisctx)
+            if ((thisctx & SSL_EXT_CLIENT_HELLO) != 0)
+                role = ENDPOINT_SERVER;
+            else if ((thisctx & SSL_EXT_TLS1_2_SERVER_HELLO) != 0)
+                role = ENDPOINT_CLIENT;
```
Oops, the duplication inside the loop is unnecessary. I've deleted it. The idea was that this would be a generic function. Although there is currently only one call site there are still two possible contexts (`SSL_EXT_TLS1_2_SERVER_HELLO` and `SSL-EXT_TLS1_3_SERVER_HELLO`). The check for `SSL_EXT_CLIENT_HELLO` could be removed at the loss of the generic nature of the function...but, I kind of feel it is better to keep it generic.

mattcaswell commented on ssl/statem/extensions.c:
```diff
@@ -361,6 +365,49 @@ static int validate_context(SSL *s, unsigned int extctx, unsigned int thisctx)
+            context = meth->context;
+        }
+
+        if (!validate_context(s, context, thisctx))
```
Exactly.

mattcaswell commented on ssl/statem/extensions_clnt.c:
```diff
@@ -482,6 +482,18 @@ EXT_RETURN tls_construct_ctos_supported_versions(SSL *s, WPACKET *pkt,
+    /* Don't include this if we can't negotiate TLSv1.3. We can do a straight
+     * comparison here because we will never be called in DTLS.
+     */
+    if (max_version < TLS1_3_VERSION)
```
In my mind this extension isn't well defined if it isn't used in a TLSv1.3 context (i.e. where TLSv1.3 isn't one of the supported versions). It doesn't gain us anything to use it where that is the case so I think it is best not to.

mattcaswell commented on ssl/statem/statem_clnt.c:
```diff
@@ -1083,13 +1107,32 @@ int tls_construct_client_hello(SSL *s, WPACKET *pkt)
         sess_id_len = s->session->session_id_length;
+        if (s->version == TLS1_3_VERSION) {
+            s->tmp_session_id_len = sess_id_len;
+            memcpy(s->tmp_session_id, s->session->session_id, sess_id_len);
```
As noted in my other comment above, reusing the existing session_id field causes problems, e.g. even if a real session doesn't have a session id (because it is a ticket), it still uses a session id under the hood (it is the key into the session cache). I think it is best to keep them separate.

mattcaswell commented on ssl/statem/statem_lib.c:
```diff
@@ -1744,24 +1761,37 @@ int ssl_choose_server_version(SSL *s, CLIENTHELLO_MSG *hello, DOWNGRADE *dgrd)
-    if (version == TLS1_3_VERSION_DRAFT)
-        version = TLS1_3_VERSION;
+    origv = s->version;
+    s->version = version;
```
Yes, it's not ideal but we don't currently have a mechanism for passing back "results" from calling `tls_parse_extension` other than by storing it in the `SSL` object. Possibly we could introduce such a thing - but it kind of seems like a sledgehammer-to-crack-a-nut. That means the supported_versions extension has to store the version "somewhere" and s->version seems the sensible place. Alternatively we could introduce a new temporary variable in SSL - but that seems equally messy.

kaduk replied:
Agreed on probably equally messy to use a new temporary variable.
I wonder what it would look like to do at the start
```
origv = s->version
s->version = version
tls_parse_extension(...)
newv = s->version
s->version = origv
```
and then we would only re-write ```s->version``` (again) for explicit change, as opposed to having to back out the change if something goes wrong.
(Obviously I didn't actually flesh out what that would be; maybe it's even worse than this.)

mattcaswell replied:
Of course that would mean we have to then change it *again* if we don't encounter an error. Which strikes me as quite messy, so I'm not really sure if this is better.

mattcaswell commented on ssl/statem/statem_srvr.c:
```diff
@@ -1636,6 +1651,12 @@ static int tls_early_post_process_client_hello(SSL *s, int *pal)
 
+    if (SSL_IS_TLS13(s)) {
+        memcpy(s->tmp_session_id, s->clienthello->session_id,
+               s->clienthello->session_id_len);
```
It needs to live until we write back the ServerHello because we echo it back there.

mattcaswell commented on ssl/statem/statem_srvr.c:
```diff
@@ -2184,28 +2218,41 @@ int tls_construct_server_hello(SSL *s, WPACKET *pkt)
 
-    if ((!SSL_IS_TLS13(s)
-                && !WPACKET_sub_memcpy_u8(pkt, s->session->session_id, sl))
+    if (!WPACKET_sub_memcpy_u8(pkt, session_id, sl)
```
Hmmm...the end result looks ok AFAICT, so probably not worth worrying about.

mattcaswell commented:
Updated commits pushed. I've had to rebase because there were some *major* conflicts with master due to the error handling overhaul, which was quite messy to merge in. Hopefully I got it right. I also fixed the travis issue and the duplicated code issue. Hopefully the rebase doesn't mess your review up too much.

mattcaswell commented:
Rebased again to resolve conflicts.

Ping? It would be really nice to get this one in asap.

levitte commented on ssl/statem/extensions_clnt.c:
```diff
@@ -507,6 +507,19 @@ EXT_RETURN tls_construct_ctos_supported_versions(SSL *s, WPACKET *pkt,
+        return EXT_RETURN_FAIL;
+    }
+
+    /* Don't include this if we can't negotiate TLSv1.3. We can do a straight
```
Nit pick: `/*` on a line of its own

mattcaswell replied:
Fixed

levitte commented on ssl/statem/statem_clnt.c:
```diff
@@ -391,14 +382,13 @@ static WRITE_TRAN ossl_statem_client13_write_transition(SSL *s)
                 || s->early_data_state == SSL_EARLY_DATA_FINISHED_WRITING)
             st->hand_state = TLS_ST_PENDING_EARLY_DATA_END;
+        else if ((s->options & SSL_OP_ENABLE_MIDDLEBOX_COMPAT) != 0
+                    && s->hello_retry_request == SSL_HRR_NONE)
```
Nit pick: I know those four extra spaces are your usual style to avoid mixing up with the body following...  but in this case, it's not really needed, is it?  If you align with the wrapping parenthesis, you'll still get the separation you desire.

mattcaswell replied:
Fixed

richsalz commented:
let's get this in and then fix bugs.  +1 from me after you fix @levitte's nits.

kaduk commented on test/tls13ccstest.c:
```diff
@@ -0,0 +1,493 @@
+
+    switch (cmd) {
+    case BIO_CTRL_DUP:
+        ret = 0L;
```
nit: 'L' suffix is not needed here

mattcaswell replied:
Fixed

mattcaswell commented on ssl/ssl_locl.h:
```diff
@@ -1116,7 +1120,7 @@ struct ssl_st {
 
     /* Flag to indicate whether we should send a HelloRetryRequest or not */
-    int hello_retry_request;
+    enum {SSL_HRR_NONE, SSL_HRR_PENDING, SSL_HRR_COMPLETE} hello_retry_request;
```
I added an explicit "= 0" against the definition of `SSL_HRR_NONE`. Maybe that makes it feel a bit less like magic?

mattcaswell commented on ssl/statem/statem_clnt.c:
```diff
@@ -1047,7 +1070,8 @@ int tls_construct_client_hello(SSL *s, WPACKET *pkt)
             || !ssl_version_supported(s, sess->ssl_version)
             || !SSL_SESSION_is_resumable(sess)) {
-        if (!ssl_get_new_session(s, 0)) {
+        if (s->hello_retry_request  == SSL_HRR_NONE
```
Fixed

mattcaswell commented on util/perl/checkhandshake.pm:
```diff
@@ -153,16 +194,21 @@ sub checkhandshake($$$$)
+                                 && $extchnum == 2;
                 next if $extensions[$extloop][0] == TLSProxy::Message::MT_CLIENT_HELLO
                                  && $extchnum != $chnum;
+                next if $extensions[$extloop][0] == TLSProxy::Message::MT_SERVER_HELLO
```
The perl code has never really complied with the C coding standards. We should probably have some perl ones - but in any case I'd say that's not something to address in this PR.

mattcaswell commented on ssl/statem/statem_srvr.c:
```diff
@@ -664,6 +668,8 @@ WORK_STATE ossl_statem_server_pre_work(SSL *s, WORK_STATE wst)
 
     case TLS_ST_SW_CHANGE:
+        if (SSL_IS_TLS13(s))
+            break;
```
I'm not really sure I understand what you're getting at here. What exactly is it you want me to emphasise in the commit message? And which commit (the one that added this code was about proper handling of HRRs)?

kaduk replied:
The commit message was

```
    Remember whether we issued an HRR
    
    Make sure we remember whether we issued an HRR or not, even after we
    have finished processing it. We need this so we know whether to send a
    CCS or not later.
```

which matches up with the bulk of the commit, namely, switching ```s->hello_retry_request``` to have that enum type (and the checks of it changed accordingly).  But this commit *also* changes the server's state machine; we go from ```TLS_ST_SW_CHANGE``` to ```TLS_ST_EARLY_DATA``` for the CCS after a HRR, now, and do not setup the key block in the pre-work for ```TLS_ST_SW_CHANGE``` (which, okay, yes, it is confusing that I commented on that one specifically; sorry about that), and in the post-work we now do nothing after writing HRR+CCS, and maybe one other thing that I missed just now.

Anyway, the state machine change seems like the actual goal of the commit; the type-change for ```s->hello_retry_request``` is just a means to that end.  But the commit message (to me) emphasizes the means and doesn't mention state machine changes actually occuring at all.

mattcaswell commented:
Nits fixed, rebased and updates pushed. I just had one outstanding query on your comment about the commit message. Please can you reconfirm?

kaduk commented:
Reconfirm +1

mattcaswell commented:
Pushed. Thanks!!
